docs/guide/handle-a-pagination-on-a-custom-collection.php

```diff
@@ -0,0 +1,142 @@
+    use Doctrine\Common\Collections\Criteria;
+    use Doctrine\ORM\Tools\Pagination\Paginator as DoctrinePaginator;
+    use Doctrine\Persistence\ManagerRegistry;
+
+    class BookRepository extends ServiceEntityRepository
+    {
+        public function __construct(ManagerRegistry $registry)
+        {
+            parent::__construct($registry, Book::class);
+        }
+
+        public function getPublishedBooks(int $page = 1, int $itemsPerPage = 30): DoctrinePaginator
+        {
+            // Retrieve the custom collection and inject it into a Doctrine Paginator object
+            return new DoctrinePaginator(
+                $this->createQueryBuilder('b')
+                     ->where('b.published = :isPublished')
+                     ->setParameter('isPublished', true)
+                     ->addCriteria(
+                         Criteria::create()
+                             ->setFirstResult(($page - 1) * $itemsPerPage)
+                             ->setMaxResults($itemsPerPage)
+                     )
+            );
+        }
+    }
+}
+
+namespace App\State {
+    use ApiPlatform\Doctrine\Orm\Paginator;
+    use ApiPlatform\Metadata\Operation;
+    use ApiPlatform\State\Pagination\Pagination;
+    use ApiPlatform\State\ProviderInterface;
+    use App\Repository\BookRepository;
+
+    class BooksListProvider implements ProviderInterface
+    {
+        public function __construct(private readonly BookRepository $bookRepository, private readonly Pagination $pagination)
+        {
+        }
+
+        public function provide(Operation $operation, array $uriVariables = [], array $context = []): Paginator
+        {
+            // Retrieve the pagination parameters from the context thanks to the Pagination object
+            [$page, , $limit] = $this->pagination->getPagination($operation, $context);
+
+            // Decorates the Doctrine Paginator object to the API Platform Paginator one
+            return new Paginator($this->bookRepository->getPublishedBooks($page, $limit));
+        }
+    }
+}
+
+namespace App\Playground {
+    use App\Kernel;
+    use Symfony\Component\HttpFoundation\Request;
+
+    function request(): Request
+    {
+        return Request::create('/books.jsonld', 'GET');
+    }
+
+    function setup(Kernel $kernel): void
+    {
+        $kernel->executeMigration();
+        $kernel->loadFixtures();
+    }
+}
+
+namespace DoctrineMigrations {
+    use Doctrine\Migrations\AbstractMigration;
+
+    final class Migration extends AbstractMigration
+    {
+        public function getDescription(): string
+        {
+            return 'Creates Book objects.';
+        }
+
+        public function up(): void
+        {
+            $this->addSql('CREATE TABLE book (id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, title VARCHAR(255) NOT NULL, is_published SMALLINT NOT NULL)');
+        }
+
+        public function down(): void
+        {
+            $this->addSql('DROP TABLE book');
+        }
```

Comment:
best practice is to only use `up`, to reduce the code here can you remove `down` and `getDescription`?